### PR TITLE
Adds --describe option to brew bundle dump

### LINKED ISSUE #322: Optionally output package description as comment
I'd like to have a Brewfile that has only the formulae that I've intended to install, no dependencies. Since Homebrew doesn't seem to track† if a formula was installed manually or as a dependency, I'm left to manually edit the output of `brew bundle dump`. It's a slog because I've got more than 350 packages installed but I'll deal.

What could really expedite my discovery of what some formulae are is including as a comment on the line or above it in the Brewfile output. That way, I don't have to do `brew info packagename` for every package I don't recognize, which is a lot since my Homebrew installation dates back a few years. Honestly, I've found that every Brewfile I've manually created for the last couple of months – something I've been dropping into every repo I touch at work – has had a comment describing what the package is or does.

I'm willing to put the time into developing this feature. I envision it as an addition [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L123)
 and [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L32). It would be enabled through a command line option to the `dump` command, probably something like `--comment` or `--describe`. There's some complexity in that it appears that there are not currently arguments passed down into the dumper.

_† One thing I noted is that maybe this assumption is out of date, given what I see [here](https://github.com/Homebrew/homebrew-bundle/blob/master/lib/bundle/brew_dumper.rb#L25): `f[:installed_on_request?] || !f[:installed_as_dependency?]`_

### PR DESCRIPTION
**Pull request by colindean**
This option outputs above each dependency a comment containing a description when the dependency has a description associated with it.

Right now, only Homebrew formulae have descriptions. Casks, taps, and mas dependencies do not have descriptions available.

Fixes #322

### CODE REVIEW

**Comment by colindean** [comment]
Working on a test for this.

File: spec/brew_dumper_spec.rb [hunk 1]
```diff
@@ -477,6 +480,99 @@
     end
   end
 
+  context "when --describe is not set" do
+    before do
+      ARGV.delete("--describe") if ARGV.include?("--describe")
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: Use `stub_const("ARGV", [])` instead.

File: spec/brew_dumper_spec.rb [hunk 2]
```diff
@@ -477,6 +480,99 @@
     end
   end
 
+  context "when --describe is not set" do
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: I wonder if this case is needed just to test `#` isn't added?

File: spec/brew_dumper_spec.rb [hunk 3]
```diff
@@ -477,6 +480,99 @@
...
+      ]
+    end
+    it "does not output a comment with dependency description" do
+      expect(Bundle::BrewDumper.dump).not_to include("#")
+    end
+  end
+
+  context "when --describe is set" do
+    before do
+      ARGV << "--describe" unless ARGV.include?("--describe")
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: Use `stub_const("ARGV", ["--describe"])` instead

File: spec/brew_dumper_spec.rb [hunk 4]
```diff
@@ -477,6 +480,99 @@
...
+          build_dependencies: [],
+          requirements: [],
+          conflicts_with: [],
+          pinned?: false,
+          outdated?: false,
+        },
+      ]
+    end
+    after do
+      ARGV.delete("--describe") if ARGV.include?("--describe")
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: Can remove this with the `stub` addition.

File: spec/brew_dumper_spec.rb [hunk 5]
```diff
@@ -477,6 +480,99 @@
...
+          optional_dependencies: [],
+          build_dependencies: [],
+          requirements: [],
+          conflicts_with: [],
+          pinned?: false,
+          outdated?: false,
+        },
+      ]
+    end
+    it "does not output a comment with dependency description" do
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: Add a newline before each `it`

File: spec/brew_dumper_spec.rb [hunk 6]
```diff
@@ -477,6 +480,99 @@
...
+          conflicts_with: [],
+          pinned?: false,
+          outdated?: false,
+        },
+      ]
+    end
+    after do
+      ARGV.delete("--describe") if ARGV.include?("--describe")
+    end
+    it "outputs a comment on the line before a dependency with a description" do
```
Comment (MikeMcQuaid) [inline spec/brew_dumper_spec.rb]: Add a newline before each `it`

File: lib/bundle/brew_dumper.rb [hunk 7]
```diff
@@ -25,7 +25,9 @@ def dump
         f[:installed_on_request?] || !f[:installed_as_dependency?]
       end
       requested_formula.map do |f|
-        brewline = "brew \"#{f[:full_name]}\""
+        brewline = ""
+        brewline += "# #{f[:desc]}\n" if ARGV.include?("--describe") && f[:desc]
```
Comment (MikeMcQuaid) [inline lib/bundle/brew_dumper.rb]: Do you think it's preferably to display this comment above vs. on the same line to the right of the name? Don't feel strongly but figured it's worth thinking about.

**Comment by MikeMcQuaid** [comment]
Thanks for your work here, @colindean!

File: lib/bundle/brew_dumper.rb (hunk 7 above)
Comment (colindean) [reply lib/bundle/brew_dumper.rb]: Above felt better for reading.

```
# this is some description
brew 'asdf'
# a tool for describing
brew 'sdfgsdfsdfdadfs'
# yet another great tool everyone installs
brew 'wertwert'
```

versus

```
brew 'asdf' # this is some description
brew 'sdfgsdfsdfdadfs' # a tool for describing
brew 'wertwert' # yet another great tool everyone installs
```

The latter is harder to skim by a lot, even when syntax highlighting is present.

File: spec/brew_dumper_spec.rb (hunk 2 above)
Comment (colindean) [reply spec/brew_dumper_spec.rb]: Eh, it was a TTD test that I left in. It's ensuring that there is no description when the user hasn't requested it. I don't see the harm in leaving it in other than that it extends tests by about 0.7 seconds (`time rspec spec/brew_dumper_spec.rb`).

**Comment by colindean** [comment]
Easy comments addressed. Let me know what you'd like for that one test.

File: lib/bundle/brew_dumper.rb (hunk 7 above)
Comment (MikeMcQuaid) [reply lib/bundle/brew_dumper.rb]: The only thought reading that is you may want an additional newline after every described line to improve the grouping?

File: spec/brew_dumper_spec.rb (hunk 2 above)
Comment (MikeMcQuaid) [reply spec/brew_dumper_spec.rb]: Ok, cool 👍

File: lib/bundle/brew_dumper.rb (hunk 7 above)
Comment (colindean) [reply lib/bundle/brew_dumper.rb]: I'd thought about that. I played with it a little. I can't decide on _where_ I want to put the newlines. I can insert this into `brew_dumper.rb` on line 35 as of the latest commit and I get reasonable output.

```ruby
brewline += "\n" if ARGV.include?("--describe") && f[:desc]
```

If I omit the `&& f[:desc]`, then _all_ brew lines get an extra newline, regardless of if they have a  description. This looks more consistent.

Honestly, the inconsistency is what makes me not want to have the extra newline.

**Comment by colindean** [comment]
Fixed the style error.

**Comment by MikeMcQuaid** [comment]
Thanks again @colindean!

**Comment by colindean** [comment]
🎉

**Pull request merged**